### PR TITLE
Refactor debug query processor output

### PR DESCRIPTION
**Pull request by metamben**
Generalize the processing of debug events in the debug query processor.

Currently, debug events include transformations made by middleware and
errors occurring during processing.

This change enables processing the events in differently from just printing,
one frequent use case being sending them to tap listeners. For this, `tap>`
can be provided as `:printer`. Together with #25377, this can be used to
display debug events in a portal browser.

With `dev.portal` loaded, `dev.debug-qp/process-query-debug` can be
made to log in portal:

```clojure
  (dev.debug-qp/process-query-debug
   (mt/mbql-query venues
     {:source-query {:source-table $$venues}
      :breakout     [[:field [:field "category_id" {:base-type :type/Integer}] nil]]
      :limit        10})
   :printer (fn [[_ middleware-var before after :as event]]
              (if (and (var? middleware-var) before after)
                (dev.portal/send-log
                 (with-meta [before after]
                   {:portal.viewer/default :portal.viewer/diff})
                 (update (meta middleware-var) :ns #(.name %)))
                (dev.portal/send-log event))))
```

### CODE REVIEW

File: dev/src/dev/debug_qp.clj [hunk 1]
```diff
@@ -193,15 +193,75 @@
...
+
+  This is the default printer of `process-query-debug`."
+  first)
+
+(defmethod print-formatted-event ::transformed-query
+  [[_ middleware-var before after]]
+  (println (format "[pre] %s transformed query:" middleware-var))
+  (when *print-full?*
+    (println (u/pprint-to-str 'cyan (format-output after))))
+  (print-diff before after))
```
Comment (metamben) [inline dev/src/dev/debug_qp.clj:208]: Could be extracted into a function.

File: dev/src/dev/debug_qp.clj [hunk 2]
```diff
@@ -193,15 +193,75 @@
...
+
+(defmethod print-formatted-event ::pre-process-query-error
+  [[_ middleware-var e]]
+  (println (format "Error pre-processing query in %s:\n%s"
+                   middleware-var
+                   (u/pprint-to-str 'red (Throwable->map e)))))
```
Comment (metamben) [inline dev/src/dev/debug_qp.clj:214]: Could be extracted into a function.

**Comment by escherize** [comment]
It would be good to get that printer function from your PR comment def'ed, maybe in the portal namespace? Also some docs about how to set up portal with this would be cool. (Looks like `(set! *printer* dev.portal/qp-printer)` or something close, right?) I can just imagine pointing people to this PR when they try setting it up, vs. sending them to the `dev.portal` ns.

**Review by escherize — APPROVED** [review]

**Comment by metamben** [comment]
I'll make a PR about the example in the description.

**Pull request merged**